### PR TITLE
[WIP] add conda env file

### PR DESCRIPTION
**Pull request by agramfort**
it would be nice to test this file in our CIs 

then we would just recommend to do:

conda env create -f environment.yml

to install a working MNE environment.

EDIT: DO NOT MERGE, need to figure out how to fix `traitsui` so we don't need to skip those tests.

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  - nibabel
+  - neo
+  - pillow
+  - nose
+  - pytest
+  - pytest-cov
+  - sphinx
+  - sphinx_bootstrap_theme
+  - numpydoc
+  - sphinx-gallery
```
Comment (larsoner) [inline environment.yml]: also needs `nilearn`

I can push and make Travis use this sometime today hopefully

Actually we should make `ci-helpers` support this

**Comment by larsoner** [comment]
See https://github.com/astropy/ci-helpers/issues/232

**Comment by larsoner** [comment]
... and now https://github.com/astropy/ci-helpers/pull/233

**Comment by larsoner** [comment]
@agramfort the current version doesn't work because of conflicts.

Even doing:
```
conda create -n test2 python=3.6
source activate test2
conda install -c menpo menpo numpy scipy matplotlib mayavi
```
I get the conflict. Can you get one to work?

I ask because once we have a working env we can probably do `conda env export` to get the right `yml` file, or at least something close.

If you have Py3k + Mayavi in an env, can you try the command to see what the yml file looks like, maybe paste it somewhere?

**Comment by kambysese** [comment]
Question about this approach. Is the recommendation to install mne-python stable or development version in the virtualenv?
If it helps at all I have a 2.7 env file that works without any of the backend or Mayavi excitement. But I see you guys are transitioning to 3.6.

**Comment by larsoner** [comment]
Stable is the recommended version for users

Hold off on sharing the file for now, I want to see if we can get 3.6 to work

**Comment by kambysese** [comment]
Thanks, @larsoner FUQ do you know if it's possible to use different dev branches across different virtual envs?

**Comment by larsoner** [comment]
Not that I know of when using a single directory, but you could always `git clone` to two different directories like and `python setup.py develop` from each of them as desired

**Comment by agramfort** [comment]
I just pushed a working env for osx

can someone try it on linux?

**Comment by larsoner** [comment]
Yes I can try today. https://github.com/astropy/ci-helpers/pull/233 is merged so I can then try to get a Travis build running with it.

**Comment by agramfort** [comment]
yes !

**Comment by larsoner** [comment]
Other than the hilarious error `ModuleNotFoundError: No module named 'mne'` the env works. I'll try making Travis use it

**Comment by agramfort** [comment]
great thanks

yeah I did not include it for travis :)

**Comment by larsoner** [comment]
> yeah I did not include it for travis :)

The `python setup.py develop` should take precedence I think. We'll find out...

File: appveyor.yml [hunk 2]
```diff
@@ -1,8 +1,7 @@
 environment:
   global:
       PYTHON: "C:\\conda"
-      CONDA_DEPENDENCIES: "setuptools numpy scipy matplotlib scikit-learn mayavi pandas h5py PIL pyside pytest"
-      PIP_DEPENDENCIES: "git+git://github.com/nipy/PySurfer.git nibabel nitime nose-timer nose-faulthandler pytest-attrib"
```
Comment (agramfort) [inline appveyor.yml]: we don't need nitime nose-timer nose-faulthandler anymore?

Comment (larsoner) [reply appveyor.yml]: we don't use `nose` anymore

**Comment by larsoner** [comment]
had to `rebase` to `master` to resolve conflicts

**Comment by larsoner** [comment]
The GUI stuff doesn't work because of a Py3k bug in `traitsui`, but when I update it to use `pip` it doesn't fix it, and when I update to git master, it breaks in a different way. So we're a bit stuck.

In other words, `pytest mne/gui/` fails, so @agramfort you have a broken env :(

**Comment by larsoner** [comment]
(but other than that, this should be working at this point)

**Comment by larsoner** [comment]
Pushed a commit that should skip those tests (which is not good) to at least see if we can find green

File: mne/gui/tests/test_kit2fiff_gui.py [hunk 3]
```diff
@@ -119,11 +120,12 @@ def test_kit2fiff_model():
...
 def test_kit2fiff_gui():
     """Test Kit2Fiff GUI."""
-    if os.environ.get('TRAVIS_OS_NAME') == 'linux':
-        raise SkipTest("Skipping on Travis for Linux due to GUI error")
     home_dir = _TempDir()
     os.environ['_MNE_GUI_TESTING_MODE'] = 'true'
     os.environ['_MNE_FAKE_HOME_DIR'] = home_dir
+    if sys.version[0] == '3' and (os.environ.get('APPVEYOR') == 'True' or
+                                  os.environ.get('TRAVIS') == 'true'):
+        raise SkipTest('traitsui broken on Py3k')
```
Comment (larsoner) [inline mne/gui/tests/test_kit2fiff_gui.py]: @agramfort this is one test that we should reenable

File: mne/gui/tests/test_coreg_gui.py [hunk 4]
```diff
@@ -146,6 +148,9 @@ def test_coreg_gui():
     home_dir = _TempDir()
     os.environ['_MNE_GUI_TESTING_MODE'] = 'true'
     os.environ['_MNE_FAKE_HOME_DIR'] = home_dir
+    if sys.version[0] == '3' and (os.environ.get('APPVEYOR') == 'True' or
+                                  os.environ.get('TRAVIS') == 'true'):
+        raise SkipTest('traitsui broken on Py3k')
```
Comment (larsoner) [inline mne/gui/tests/test_coreg_gui.py]: And this is the other.

**Comment by agramfort** [comment]
@larsoner I would open an issue on traitsui saying that we need qt4 with traitsui for mayavi that does not support yet qt5.

I hacked it by chaning L433 of enum_editor.py to:

name = names[min(int(index), len(names)) - 1]

but it's ugly.

@christianbrodbeck do you have an idea of how to go around the bug?

**Comment by larsoner** [comment]
It looks like a float indexing problem (look at the error log and you'll
see it), on their master they cast to int so I assumed it was fixed and
thus a packaging problem (using git master broke things a different way)

**Comment by larsoner** [comment]
i.e., I think the only change you need is the `int()` and it's already in their git `master`, but when I tried using `pip` to install it from GitHub it broke some other package

**Comment by agramfort** [comment]
here is a 2.7 env that seems to work:

```
name: mne27
channels:
- conda-forge
- defaults
dependencies:
- python=2.7.12=1
- pip
- mkl
- numpy
- scipy
- pandas
- scikit-learn
- jupyter
- ipython
- statsmodels
- h5py
- pyqt=4.11.4
- qt=4.8.7
- tk=8.5.18=0
- traits=4.4.0=py27_0
- traitsui=4.4.0=py27_0
- traitlets=4.3.2=py27_0
- vtk=5.10.1=py27_0
- mayavi=4.4.0=np19py27_0
- matplotlib
- sphinx
- nose
- pytest
- pip:
  - mne
  - "git+https://github.com/nipy/PySurfer.git"
  - nibabel
  - nilearn
  - neo
  - pillow
  - pytest-cov
  - pytest-sugar
  - sphinx_bootstrap_theme
  - numpydoc
  - sphinx-gallery
```

take home message is that traits* 4.4 works with our tests on qt4 but 4.4 is not available on conda with python 3k ... oh well ...

**Comment by larsoner** [comment]
So let's stick with 2.7 for now and keep the 3.6 around (we can even still test it) in case people want to use it. I'll give it a shot soon

**Comment by agramfort** [comment]
sounds good

the "real" fix would be to migrate mayavi to qt5 hum...

**Comment by larsoner** [comment]
From [here](http://docs.enthought.com/mayavi/mayavi/bugs.html):

> Qt 5: Mayavi is built on top of the traitsui and pyface gui libraries. Thus support for Qt 5 is linked to these gui libraries supporting Qt 5.

**Comment by agramfort** [comment]
hum... is there any open bug with discussion on traitsui tracker?

**Comment by larsoner** [comment]
http://lmgtfy.com/?q=traitsui+qt5

**Comment by larsoner** [comment]
Appently as of 11 days ago (at least) they have it via `pyface`:

https://github.com/enthought/traitsui/pull/242#issuecomment-324620957

**Comment by agramfort** [comment]
haha :)

**Comment by larsoner** [comment]
So I almost got 3.6 working properly with Qt5, the blocker is now PySide, which does not work on 3.6:

https://github.com/pyside/PySide/issues/132

:(

**Comment by larsoner** [comment]
I'll see if I can get Qt5 working in the 2.7 env, though

**Comment by mwaskom** [comment]
I wish there were some way to force all the people who obnoxiously declare that there's "no reason to stay on Python 2" to read through all these issue chains.

**Comment by larsoner** [comment]
Yeah no kidding.

I also spent some time trying to get Qt5 on the 2.7 env and that failed. But it looks like the Qt4/2.7 should work, we'll see.

**Comment by agramfort** [comment]
@mwaskom what a surprise to see you here :)

yes 2.7 and qt4 should work with the env I gave you.

**Comment by larsoner** [comment]
Keeping track of `sklearn` versions and API changes is a [bit of a nightmare](https://ci.appveyor.com/project/mne-tools/mne-python/build/1.0.495)

**Comment by larsoner** [comment]
@agramfort the `environment2.yml` seems to have a problem on Windows -- it installs `sklearn` 0.16, which means most new stuff in `decoding` can't be used :(

It also just crashes (without any output!) on Travis. I'm a bit stuck.

**Comment by larsoner** [comment]
Apparently `environment3.yml` has the same problem on Windows

**Comment by larsoner** [comment]
... but when I use `conda env create -f environment3.yml` on my Windows VM it uses sklearn 0.19. I don't get it.

**Comment by larsoner** [comment]
Scratch that, it's just the 2.7 env that has sklearn 0.16, the 3.6 env has 0.19

**Comment by larsoner** [comment]
There is https://github.com/enthought/mayavi/pull/528, so maybe we have a chance at PyQt5 (which was holding back `matplotlib`)

**Comment by agramfort** [comment]
@larsoner can you split the pytest migration and the datetime fix to other PRs to help us review.

then it will "just" be changes in CIs scripts.

thanks a lot

**Comment by larsoner** [comment]
> can you split the pytest migration and the datetime fix to other PRs to help us review.

Sure, hopefully this weekend but maybe not until Monday

**Comment by larsoner** [comment]
@agramfort I'll start a new PR once the other changes are merged.

**Pull request closed without merging**